File: decl-ts/tests/fmt_test.ts
// The formatter: the canonical-form cases (tests/fmt/cases.json), then its
// two properties over every parseable module of the corpora — idempotent
// (fmt(fmt(x)) == fmt(x)) and AST-preserving (formatting moves columns,
// never nodes).
import { readFileSync } from 'node:fs';
import { join } from 'node:path';
import { initFormatter, format } from '../src/fmt.ts';
import { parseSource } from '../src/parse.ts';
import { initParser } from '../src/node.ts';
import { walkDecl } from '../src/conformance.ts';
import { check, total, root } from './common/check.ts';

await initParser();
await initFormatter();

type Case = { name: string; input: string; expected?: string; error?: boolean };

console.log('== fmt: the cases of tests/fmt ==');
const cases: Case[] = JSON.parse(readFileSync(join(root, 'tests/fmt/cases.json'), 'utf8'));
for (const c of cases) {
  let got: string | null = null,
    threw = '';
  try {
    got = format(c.input);
  } catch (e: any) {
    threw = e.message;
  }
  if (c.error) check(c.name, got === null, `formatted anyway: ${JSON.stringify(got)}`);
  else
    check(
      c.name,
      got === c.expected,
      JSON.stringify({ got: got ?? `THROW ${threw}`, want: c.expected }),
    );
}

console.log('== fmt: idempotent and AST-preserving over the corpora ==');
{
  const files: string[] = [];
  for (const dir of ['tests/validation', 'tests/modules', 'tests/packages', 'docs/examples'])
    for (const f of walkDecl(join(root, dir))) files.push(f);
  let idem = 0,
    skipped = 0,
    idemFail = 0,
    tokenFail = 0;
  const tokens = (src: string) =>
    JSON.stringify(parseSource(src).decls, (k, v) =>
      k === 'loc' ? undefined : typeof v === 'bigint' ? `${v}n` : v,
    ); // source ranges are not the AST
  for (const f of files) {
    const src = readFileSync(f, 'utf8');
    if (parseSource(src).errors.length) {
      skipped++;
      continue;
    } // invalid-parsing fixtures
    let once: string, twice: string;
    try {
      once = format(src);
    } catch {
      skipped++;
      continue;
    }
    try {
      twice = format(once);
    } catch (e: any) {
      idemFail++;
      console.log(`  SECOND PASS FAILS ${f.slice(root.length + 1)}: ${e.message}`);
      continue;
    }
    if (once === twice) idem++;
    else {
      idemFail++;
      console.log(`  NOT IDEMPOTENT ${f.slice(root.length + 1)}`);
    }
    if (parseSource(once).errors.length !== 0 || tokens(once) !== tokens(src)) {
      tokenFail++;
      console.log(`  AST CHANGED ${f.slice(root.length + 1)}`);
    }
  }
  check(
    `fmt(fmt(x)) == fmt(x) on ${idem + idemFail} parseable files`,
    idemFail === 0,
    `${idemFail} failures`,
  );
  check(`formatting preserves the AST on all files`, tokenFail === 0, `${tokenFail} failures`);
  console.log(`  (${skipped} unparseable fixtures skipped by design)`);
}
total();

File: decl-rs/tests/fmt_test.rs
//! The formatter: the canonical-form cases (tests/fmt/cases.json), then its
//! two properties over every parseable module of the corpora — idempotent
//! (fmt(fmt(x)) == fmt(x)) and AST-preserving (formatting moves columns,
//! never nodes).
mod common;
use common::*;
use decl_lang::conformance::walk_decl;
use decl_lang::fmt::format;
use decl_lang::parse::parse_source;

#[test]
fn fmt_corpus() {
    let text = std::fs::read_to_string(root().join("tests/fmt/cases.json")).unwrap();
    let Ok(Value::JArr(cases)) = read_json(&text) else {
        panic!("cases.json is a list")
    };
    let mut failures: Vec<String> = vec![];
    for c in cases.iter() {
        let name = common::text(get(c, "name").unwrap());
        let got = format(common::text(get(c, "input").unwrap()));
        match (get(c, "error"), get(c, "expected")) {
            (Some(Value::Bool(true)), _) => {
                if let Ok(s) = &got {
                    failures.push(format!("{name}: formatted anyway: {s:?}"));
                }
            }
            (_, Some(want)) => {
                let want = common::text(want);
                if got.as_deref() != Ok(want) {
                    failures.push(format!(
                        "{name}\n      expected {want:?}\n      got      {got:?}"
                    ));
                }
            }
            _ => failures.push(format!("{name}: a case needs expected or error")),
        }
    }
    println!("fmt corpus: {} cases", cases.len());
    assert!(
        failures.is_empty(),
        "fmt corpus failures:\n  {}",
        failures.join("\n  ")
    );
}

#[test]
fn fmt_idempotent_and_ast_preserving_over_the_corpora() {
    let mut files = vec![];
    for d in [
        "tests/validation",
        "tests/modules",
        "tests/packages",
        "docs/examples",
    ] {
        walk_decl(&root().join(d), &mut files);
    }
    // the AST without its source ranges (formatting moves nodes; it must not change them):
    // every node's `loc` is blanked in the debug rendering
    let loc_re = regex::Regex::new(r"loc: (None|Some\(Loc \{ [^}]* \}\))").unwrap();
    let tokens = |src: &str| {
        let ds = parse_source(src).decls;
        loc_re.replace_all(&format!("{ds:?}"), "loc: _").to_string()
    };
    let (mut idem, mut idem_fail, mut token_fail, mut skipped) = (0, 0, 0, 0);
    for f in files {
        let src = std::fs::read_to_string(&f).unwrap();
        if !parse_source(&src).errors.is_empty() {
            skipped += 1;
            continue;
        }
        let Ok(once) = format(&src) else {
            skipped += 1;
            continue;
        };
        let twice = match format(&once) {
            Ok(t) => t,
            Err(e) => {
                idem_fail += 1;
                eprintln!("SECOND PASS FAILS {}: {e}", f.display());
                continue;
            }
        };
        if once == twice {
            idem += 1
        } else {
            idem_fail += 1;
            eprintln!("NOT IDEMPOTENT {}", f.display());
        }
        if !(parse_source(&once).errors.is_empty() && tokens(&once) == tokens(&src)) {
            token_fail += 1;
            eprintln!("AST CHANGED {}", f.display());
        }
    }
    assert!(
        idem_fail == 0,
        "fmt(fmt(x)) == fmt(x) on {} parseable files",
        idem + idem_fail
    );
    assert!(token_fail == 0, "formatting preserves the AST on all files");
    eprintln!("({skipped} unparseable fixtures skipped by design)");
}
